### PR TITLE
created: visualization evaluating assembly - IGV

### PR DESCRIPTION
**Pull request by evezeyl**
DO NOT MERGE YET: 
- includes premapping and gap and some scripts
draft now

### CODE REVIEW

**Comment by evezeyl** [comment]
Do not merge yet! started to put the outline, need to continue and to test:
- mapping reads to assembly -> preparing to visualize with IGV
- installing and visualizing with IGV

File: assembly_visualization.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
+# Visualize your assembly with [IGV](http://software.broadinstitute.org/software/igv/home)
+
+ **IGV**: Integrative Genomics Viewer.
+
+ We will use this software to evaluate our assemblies, among others to view how the reads map to the assembly.
+
+ We will also add tracks with genome annotation and gaps.
+
+We have assembled reads and annotated using Bifrost pipeline. The data for today's practical are found in `/projects/nn9305k/Bioinfo_mapping_training`
+
```
Comment (karinlag) [inline assembly_visualization.md]: Move under course folder

File: assembly_visualization.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+## 1. Mapping the reads against a reference/assembly
+
+### 1.1 What is read-mapping
+
+It is simply finding a/the matching locus/area of a read on a sequence. Usually few mismatches are allowed (think about the consequences).
```
Comment (karinlag) [inline assembly_visualization.md]: define what a matching region is

File: assembly_visualization.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+Reads can be mapped as paired or single. If paired is used, then the matching regions are defined by the insert size and the length of each read
+
+<center> <a href="https://commons.wikimedia.org/wiki/File:Mapping_Reads.png"> <img src="https://upload.wikimedia.org/wikipedia/commons/2/2e/Mapping_Reads.png" width="400">
+<br>
+
+
+### 1.2 why mapping reads
+
```
Comment (karinlag) [inline assembly_visualization.md]: Correct formatting issues.

Comment (karinlag) [reply assembly_visualization.md]: Also add Intro statement

File: assembly_visualization.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+2) to identify SNPs: some methods use reads mapping against a reference to identify and type SNPs
+
+3) Assembly polishing software such as `pilon` and `reapr` use mapped reads to identify potential improvement in assemblies. It is good to visualize what information they are using ...
+  > eg. variation in coverage, wrong read pairs orientation, discrepancy between expected insert size and actual insert size obtained from read mapping (ie. longer than expected)
+
+ - [ ] other?
+
+### 1.3 Re-map the reads to the assembly
+
+We use [bwa mem](http://bio-bwa.sourceforge.net/bwa.shtml) option from [bwa tools](http://bio-bwa.sourceforge.net/) software.
```
Comment (karinlag) [inline assembly_visualization.md]: Correct issues with links

File: assembly_visualization.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+You will use the trimmed-filtered reads in: `/projects/nn9305k/Bioinfo_mapping_training/bifrost_run/bbduk_trimmed` and the assembly: `/projects/nn9305k/Bioinfo_mapping_training/bifrost_run/prokka/MiSeq_Ecoli_MG1655_50x.fna`
+
```
Comment (karinlag) [inline assembly_visualization.md]: Which assembly are we actually using? Pilon or prokka.

File: assembly_visualization.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+We could use the assembly output from `Pilon` **BUT** then we would have to create aliases between labels used by `Prokka` and by `Pilon` to be able to visualize the assembly with IGV. `Prokka` assembly is identical to `Pilon` assembly, only labels have been changed.
```
Comment (karinlag) [inline assembly_visualization.md]: Did not understand :)

File: assembly_visualization.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+In your home make a directory for today's work
+and a folder called `mapping` where you will copy the input files
+
+```bash
+# softlink the required input files in the mapping folder:
+ln -s <path>/bifrost_run/bbduk_trimmed/*.fq.gz .
```
Comment (karinlag) [inline assembly_visualization.md]: Copy files, not softlink.

File: assembly_visualization.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+ln -s <path>/bifrost_run/pilon/MiSeq_Ecoli_MG1655_50x_pilon_spades.fasta .
+
+# we will use Bifrost with the command line
+# As bwa and samtools are used to map reads for Pilon ...
+
+qlogin --account=nn9305k --time=00:30:00 --ntasks=4 --mem-per-cpu=4G
+source activate bifrost
+
+# 1) We need to index the reference
```
Comment (karinlag) [inline assembly_visualization.md]: What is the reference sequence. 

WE are in this case using the assembly as a reference sequence.

File: assembly_visualization.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+# We need to merge those two files as one
+samtools merge <out:all_merged.bam> <in1:_.bam> <in2:.bam>
+
+# To be sure reads are still sorted: we resort
+samtools sort -o <out:final_all_merged.bam> <in:all_merged.bam>
+
+#NB:Some software like Pilon need that we update the index (you can do:). We will need index after
+samtools index <final_all_merged.bam>
+```
```
Comment (karinlag) [inline assembly_visualization.md]: Put this part into a OPTIONAL box, with explanation.

File: assembly_visualization.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+In your home make a directory for today's work
+and a folder called `mapping` where you will copy the input files
+
+```bash
+# softlink the required input files in the mapping folder:
+ln -s <path>/bifrost_run/bbduk_trimmed/*.fq.gz .
+ln -s <path>/bifrost_run/pilon/MiSeq_Ecoli_MG1655_50x_pilon_spades.fasta .
+
+# we will use Bifrost with the command line
```
Comment (karinlag) [inline assembly_visualization.md]: We will use the software included in the Bifrost pipeline. This is available as a conda environment, called bifrost.

File: assembly_visualization.md (hunk 7 above)
Comment (karinlag) [reply assembly_visualization.md]: add step where people do head on input files.

File: assembly_visualization.md [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+ `less <filename.sam>`
+
+### 1.5 looking at how the reads maps against the reference with [samtools](http://www.htslib.org/doc/samtools.html) pileup module
+
+```bash
+# To see the list of nodes:
+grep "^>" <reference.fasta>
+# look at the pileup in one node
+samtools mpileup <in:final_all_merged.bam> --max-depth 0 --fasta-ref <reference.fasta> -r <ex:NODE_76_length_637_cov_0.071429_pilon> | less
```
Comment (karinlag) [inline assembly_visualization.md]: samtools tview bamfile reference.fasta

File: assembly_visualization.md [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+
+### 1.3 Re-map the reads to the assembly
+
+We use [bwa mem](http://bio-bwa.sourceforge.net/bwa.shtml) option from [bwa tools](http://bio-bwa.sourceforge.net/) software.
+
+You will use the trimmed-filtered reads in: `/projects/nn9305k/Bioinfo_mapping_training/bifrost_run/bbduk_trimmed` and the assembly: `/projects/nn9305k/Bioinfo_mapping_training/bifrost_run/prokka/MiSeq_Ecoli_MG1655_50x.fna`
+
+We could use the assembly output from `Pilon` **BUT** then we would have to create aliases between labels used by `Prokka` and by `Pilon` to be able to visualize the assembly with IGV. `Prokka` assembly is identical to `Pilon` assembly, only labels have been changed.
+
+In your home make a directory for today's work
```
Comment (karinlag) [inline assembly_visualization.md]: project home directory

File: assembly_visualization.md [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+Have a look at:
+- coverage
+- gaps positions
+- some strange scaffolds?
+- PE orientations: in detail how the reads map to your assembly (you will need to zoom a lot)
+- are some PE reads miss-oriented? reported as having abnormal insert sizes?
+
+---------------------------------------------------------------------
+# 3 Additional: assembly evaluation
```
Comment (karinlag) [inline assembly_visualization.md]: cut this, do it later :)

File: assembly_visualization.md (hunk 1 above)
Comment (evezeyl) [reply assembly_visualization.md]: done

File: assembly_visualization.md (hunk 2 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok, tried biological ..

File: assembly_visualization.md (hunk 4 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 5 above)
Comment (evezeyl) [reply assembly_visualization.md]: simplified

File: assembly_visualization.md (hunk 7 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 8 above)
Comment (evezeyl) [reply assembly_visualization.md]: added

File: assembly_visualization.md (hunk 10 above)
Comment (evezeyl) [reply assembly_visualization.md]: added

File: assembly_visualization.md (hunk 9 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 3 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 6 above)
Comment (evezeyl) [reply assembly_visualization.md]: simplified

File: assembly_visualization.md (hunk 12 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 11 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

File: assembly_visualization.md (hunk 13 above)
Comment (evezeyl) [reply assembly_visualization.md]: ok

**Comment by evezeyl** [comment]
Think everything is ok, will have a short look again this week end, i case missed sth,

**Comment by evezeyl** [comment]
Think its ready to merge now.
Included the last minute modifications
then we do not install biopython on pc as it is in Bifrost :)

**Pull request merged**